src/main/java/org/broadinstitute/dsde/consent/ontology/translate/service/Translate.java

```diff
@@ -0,0 +1,52 @@
+package org.broadinstitute.dsde.consent.ontology.translate.service;
+
+import org.broadinstitute.dsde.consent.ontology.Utils;
+import org.broadinstitute.dsde.consent.ontology.translate.DTO.RecommendationDTO;
+import org.broadinstitute.dsde.consent.ontology.translate.model.TermItem;
+import org.broadinstitute.dsde.consent.ontology.translate.utils.LoadUtils;
+import org.broadinstitute.dsde.consent.ontology.translate.utils.SearchUtils;
+
+import java.util.HashMap;
+import java.util.List;
+
+public class Translate {
+
+    String jsonFilePath;
+
+    public Translate(String jsonFilePath) {
+        this.jsonFilePath = jsonFilePath;
+    }
+
+    public HashMap<String, RecommendationDTO> paragraph(final String paragraph) throws Exception {
+        HashMap<String, RecommendationDTO> recommendations = new HashMap<>();
+
+        if (Utils.isNullOrEmpty(paragraph) || Utils.isNullOrEmpty(jsonFilePath)) {
+            return recommendations;
+        }
+
+        List<TermItem> terms = LoadUtils.termsFromLocalResourcesJSON(jsonFilePath);
+
+        for (TermItem term : terms) {
+            final String title = term.getTitle();
+            final String category = term.getCategory();
+            final String url = term.getUrl();
+            final String[] keywords = term.getKeywords();
+
+            for (String keyword : keywords) {
+                final boolean foundMatch = SearchUtils.searchForKeyword(keyword, paragraph);
```

Comment:
If the only way a term can be matched is via its keywords, a better way to represent this may be to make a associative array that maps every keyword to its `TermItem`s, maybe `Map<String, List<TermItem>>`. Then you could search over all words in a paragraph and for each word do a keyword lookup (which is O(1)) so the term search would be O(n) instead of O(n^2). You'd have to combine all found terms into a single list and then convert that into your DTO.

Reply:
Reverse index with keywords? Clever. I like that.